ApeWorX/ape
# feat: Raise warning if `value=` is specified for anything but a `payable` function [APE-1283]

### What I did

Raise warning if value= is specified for anything but a payable function


fixes: #1396 (Raise warning if value= is specified for anything but a payable function [APE-874])

### How I did it
Added check to see if `msg.value` is getting passed to a `non-payable` contract

### How to verify it
1. Create a contract with a non-payable constructor like this
```js
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

contract NonPayableConstructor {

    address public owner;
    constructor() {
        owner = msg.sender;
    }

    function getOwner() external view returns(address){
        return owner;
    }

}
```
3. Create a `test` or `script` with the following contents
```py
#!/usr/bin/python3
from ape import project, accounts

user = accounts[1]
user.set_autosign(True)
user.balance += int(100e18)


def test_NonPayableConstructor():
    npc = project.NonPayableConstructor.deploy(sender=user, value="1 ether")
    
    print(npc.getOwner())

def main():
    test_NonPayableConstructor()
``` 
4. run using `ape run path/to/script` or `ape test -k testName`
5. You will see an expectation like this `ERROR: (ContractError) Sending funds to a non-payable constructor!!`
![image](https://github.com/ApeWorX/ape/assets/31238298/3c47b926-efae-48f3-b48d-ebaa57102dad)


### Checklist



- [x] Documentation has been updated ~~(Only applicable if there is a page for Exceptions)~~ you added a docstring to the excpetion, that will get autogenerated
